# NLP - Natural Language Processing
## Fake News Detection Project - using TF_IDF Vector

### Import relevent libraries

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer, HashingVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
import matplotlib.pyplot as pyplot

### Data Collection

In [2]:
raw_dataset = pd.read_csv('news.csv', index_col=None)
raw_dataset

Unnamed: 0                                              title  \
0           8476                       You Can Smell Hillary’s Fear   
1          10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2           3608        Kerry to go to Paris in gesture of sympathy   
3          10142  Bernie supporters on Twitter erupt in anger ag...   
4            875   The Battle of New York: Why This Primary Matters   
...          ...                                                ...   
6330        4490  State Department says it can't find emails fro...   
6331        8062  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...   
6332        8622  Anti-Trump Protesters Are Tools of the Oligarc...   
6333        4021  In Ethiopia, Obama seeks progress on peace, se...   
6334        4330  Jeb Bush Is Suddenly Attacking Trump. Here's W...   

                                                   text label  
0     Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1     Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2     U.S. Secretary of State John F. Kerry said Mon...  REAL  
3     — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4     It's primary day in New York and front-runners...  REAL  
...                                                 ...   ...  
6330  The State Department told the Republican Natio...  REAL  
6331  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...  FAKE  
6332   Anti-Trump Protesters Are Tools of the Oligar...  FAKE  
6333  ADDIS ABABA, Ethiopia —President Obama convene...  REAL  
6334  Jeb Bush Is Suddenly Attacking Trump. Here's W...  REAL  

[6335 rows x 4 columns]

### Data Preprocessing

In [3]:
dataset = raw_dataset.drop('Unnamed: 0',axis=1)
dataset

title  \
0                          You Can Smell Hillary’s Fear   
1     Watch The Exact Moment Paul Ryan Committed Pol...   
2           Kerry to go to Paris in gesture of sympathy   
3     Bernie supporters on Twitter erupt in anger ag...   
4      The Battle of New York: Why This Primary Matters   
...                                                 ...   
6330  State Department says it can't find emails fro...   
6331  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...   
6332  Anti-Trump Protesters Are Tools of the Oligarc...   
6333  In Ethiopia, Obama seeks progress on peace, se...   
6334  Jeb Bush Is Suddenly Attacking Trump. Here's W...   

                                                   text label  
0     Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1     Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2     U.S. Secretary of State John F. Kerry said Mon...  REAL  
3     — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4     It's primary day in New York and front-runners...  REAL  
...                                                 ...   ...  
6330  The State Department told the Republican Natio...  REAL  
6331  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...  FAKE  
6332   Anti-Trump Protesters Are Tools of the Oligar...  FAKE  
6333  ADDIS ABABA, Ethiopia —President Obama convene...  REAL  
6334  Jeb Bush Is Suddenly Attacking Trump. Here's W...  REAL  

[6335 rows x 3 columns]

In [4]:
dataset.shape

(6335, 3)

### Input Output Split

In [5]:
indep_x = dataset['text']
dep_y = dataset['label']

### Train Test Split

In [6]:
x_train,x_test,y_train,y_test = train_test_split(indep_x,dep_y,test_size = 0.2, random_state = 0)

## Using Counter Vector Method
### Converting Text --> Vectors (Numbers) 

In [7]:
tfidf_vector = TfidfVectorizer(stop_words = 'english')
vectors_train = tfidf_vector.fit_transform(x_train)
vectors_test = tfidf_vector.transform(x_test)

In [8]:
print(x_train)

2402    Christian Whiton is a former deputy special en...
1922    Super Tuesday Brings Harsh Light And Heartache...
3475    Prev post Page 1 of 4 Next \nNurses are among ...
6197    The deadly hostage situation at a luxury hotel...
4748    Our new country: Women and minorities hit hard...
                              ...                        
4931    Twenty-two of the 37 corporations nominated fo...
3264    As pieces of luggage, human remains, wreckage ...
1653    0 Add Comment \nIN THE immediate aftermath of ...
2607    Palestine Palestinians check the flat of Amjad...
2732    For the second week in a row, there was a temp...
Name: text, Length: 5068, dtype: object


In [9]:
print(vectors_train)

  (0, 5434)	0.04558113027328849
  (0, 22112)	0.03101556231290688
  (0, 41980)	0.041224221139183156
  (0, 41399)	0.08707226417897605
  (0, 32656)	0.05826505252086948
  (0, 40634)	0.08300352400740964
  (0, 59168)	0.03304164565844389
  (0, 17072)	0.015280054882096362
  (0, 21122)	0.03467645674134631
  (0, 36500)	0.036461647252457555
  (0, 21811)	0.021815222286389203
  (0, 55510)	0.037920004934453785
  (0, 35473)	0.025244871570555313
  (0, 8579)	0.03467645674134631
  (0, 34917)	0.05447353166186797
  (0, 51186)	0.027977956004355133
  (0, 17007)	0.020253495378206955
  (0, 13362)	0.023133406428959745
  (0, 28054)	0.026346675461032608
  (0, 51479)	0.04368398576036871
  (0, 19911)	0.03226033674797147
  (0, 12812)	0.021077575874407357
  (0, 5104)	0.0369438299052932
  (0, 22586)	0.03830998824285092
  (0, 23706)	0.03360195028163217
  :	:
  (5067, 25947)	0.0690536695761875
  (5067, 20495)	0.060909276073445075
  (5067, 46270)	0.05755520504375683
  (5067, 56317)	0.0733883544907381
  (5067, 59168)	0.0

In [10]:
print(vectors_test)

  (0, 60873)	0.04199293292033408
  (0, 60756)	0.03102467179929353
  (0, 60501)	0.025828759280643834
  (0, 60332)	0.022016410662716892
  (0, 60328)	0.029457870632407204
  (0, 60325)	0.05310926248634118
  (0, 60321)	0.037769691024057916
  (0, 60309)	0.028599297592098793
  (0, 60242)	0.02078664460607839
  (0, 60238)	0.02304876060617083
  (0, 60112)	0.06661620569458904
  (0, 60053)	0.026671179085014728
  (0, 59998)	0.08675332085991751
  (0, 59769)	0.018014389086953634
  (0, 59489)	0.01857186718038564
  (0, 59463)	0.029849288660946164
  (0, 59407)	0.03540084521652464
  (0, 59298)	0.03597852101133186
  (0, 59294)	0.07902793135900811
  (0, 59273)	0.025874119034496358
  (0, 59218)	0.030026827265587194
  (0, 59168)	0.03570687883899036
  (0, 59054)	0.055031944270905395
  (0, 58993)	0.05992243785134022
  (0, 58912)	0.049809921799913816
  :	:
  (1266, 5288)	0.04448311123524479
  (1266, 4956)	0.017480678533077217
  (1266, 4712)	0.02575856859832166
  (1266, 4225)	0.027560846706771686
  (1266, 3720)	

### Check the count of features formed from the text

In [11]:
len(tfidf_vector.get_feature_names_out())

61713

In [12]:
print(vectors_train.toarray())

[[0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.05239028 0.         ... 0.         0.         0.        ]
 [0.         0.         0.         ... 0.         0.         0.        ]
 ...
 [0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.02243351 0.         ... 0.         0.         0.        ]]


In [13]:
print(vectors_test.toarray())

[[0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.02944887 0.         ... 0.         0.         0.        ]
 ...
 [0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.01694536 0.         ... 0.         0.         0.        ]]


In [14]:
vectors_train.shape

(5068, 61713)

In [15]:
vectors_test.shape

(1267, 61713)

### Model Creation

### MultinomialNB

In [16]:
classifier = MultinomialNB()
classifier.fit(vectors_train, y_train)

MultinomialNB()

### Prediction

In [17]:
y_pred = classifier.predict(vectors_test)
y_pred

array(['REAL', 'REAL', 'FAKE', ..., 'REAL', 'REAL', 'FAKE'], dtype='<U4')

### Accuracy

In [18]:
accuracy = accuracy_score(y_test,y_pred)
accuracy

0.8318863456985004

### Confusion Matrix

In [19]:
cm = confusion_matrix(y_test,y_pred)
cm

array([[420, 195],
       [ 18, 634]], dtype=int64)

### Classification Report

In [20]:
report = classification_report(y_test,y_pred)
print(report)

              precision    recall  f1-score   support

        FAKE       0.96      0.68      0.80       615
        REAL       0.76      0.97      0.86       652

    accuracy                           0.83      1267
   macro avg       0.86      0.83      0.83      1267
weighted avg       0.86      0.83      0.83      1267



### Checking for the Prediction

In [21]:
dataset['text'][0]

'Daniel Greenfield, a Shillman Journalism Fellow at the Freedom Center, is a New York writer focusing on radical Islam. \nIn the final stretch of the election, Hillary Rodham Clinton has gone to war with the FBI. \nThe word “unprecedented” has been thrown around so often this election that it ought to be retired. But it’s still unprecedented for the nominee of a major political party to go war with the FBI. \nBut that’s exactly what Hillary and her people have done. Coma patients just waking up now and watching an hour of CNN from their hospital beds would assume that FBI Director James Comey is Hillary’s opponent in this election. \nThe FBI is under attack by everyone from Obama to CNN. Hillary’s people have circulated a letter attacking Comey. There are currently more media hit pieces lambasting him than targeting Trump. It wouldn’t be too surprising if the Clintons or their allies were to start running attack ads against the FBI. \nThe FBI’s leadership is being warned that the entir

In [22]:
vectors_train[[0]]

<1x61713 sparse matrix of type '<class 'numpy.float64'>'
	with 306 stored elements in Compressed Sparse Row format>

In [23]:
x_train[[0]]

0    Daniel Greenfield, a Shillman Journalism Fello...
Name: text, dtype: object

In [24]:
classifier.predict(vectors_train[[0]])

array(['REAL'], dtype='<U4')

In [25]:
vectors_train[25]

<1x61713 sparse matrix of type '<class 'numpy.float64'>'
	with 354 stored elements in Compressed Sparse Row format>

In [26]:
x_train[[25]]

25    Washington (CNN) The faction of the GOP that i...
Name: text, dtype: object

In [27]:
classifier.predict(vectors_train[[25]])

array(['FAKE'], dtype='<U4')

### PassiveAggressiveClassifier Algorithm

In [28]:
Classifier = PassiveAggressiveClassifier(max_iter=1000, random_state=0)
Classifier.fit(vectors_train,y_train)

PassiveAggressiveClassifier(random_state=0)

In [29]:
y_pred = Classifier.predict(vectors_test)
y_pred

array(['REAL', 'FAKE', 'FAKE', ..., 'FAKE', 'REAL', 'FAKE'], dtype='<U4')

In [30]:
accuracy = accuracy_score(y_test,y_pred)
accuracy

0.9313338595106551

In [31]:
cm = confusion_matrix(y_test,y_pred)
cm

array([[568,  47],
       [ 40, 612]], dtype=int64)

In [32]:
report = classification_report(y_test,y_pred)
print(report)

              precision    recall  f1-score   support

        FAKE       0.93      0.92      0.93       615
        REAL       0.93      0.94      0.93       652

    accuracy                           0.93      1267
   macro avg       0.93      0.93      0.93      1267
weighted avg       0.93      0.93      0.93      1267



### Logistic Regression Algorithm

In [33]:
Classifier = LogisticRegression()
Classifier.fit(vectors_train,y_train)

LogisticRegression()

In [34]:
y_pred = Classifier.predict(vectors_test)
y_pred

array(['REAL', 'FAKE', 'FAKE', ..., 'FAKE', 'REAL', 'FAKE'], dtype=object)

In [35]:
accuracy = accuracy_score(y_test,y_pred)
accuracy

0.914759273875296

In [36]:
cm = confusion_matrix(y_test,y_pred)
cm

array([[570,  45],
       [ 63, 589]], dtype=int64)

In [37]:
report = classification_report(y_test,y_pred)
print(report)

              precision    recall  f1-score   support

        FAKE       0.90      0.93      0.91       615
        REAL       0.93      0.90      0.92       652

    accuracy                           0.91      1267
   macro avg       0.91      0.92      0.91      1267
weighted avg       0.92      0.91      0.91      1267



### KNN Algorithm

In [38]:
Classifier = KNeighborsClassifier()
Classifier.fit(vectors_train,y_train)

KNeighborsClassifier()

In [39]:
y_pred = Classifier.predict(vectors_test)
y_pred

array(['FAKE', 'FAKE', 'FAKE', ..., 'FAKE', 'FAKE', 'FAKE'], dtype=object)

In [40]:
accuracy = accuracy_score(y_test,y_pred)

cm = confusion_matrix(y_test,y_pred)

report = classification_report(y_test,y_pred)
print('Accuracy:', accuracy, '\n',
      'Confusion Matrix:\n', cm, '\n',
      'Report:\n', report)

Accuracy: 0.5611681136543015 
 Confusion Matrix:
 [[615   0]
 [556  96]] 
 Report:
               precision    recall  f1-score   support

        FAKE       0.53      1.00      0.69       615
        REAL       1.00      0.15      0.26       652

    accuracy                           0.56      1267
   macro avg       0.76      0.57      0.47      1267
weighted avg       0.77      0.56      0.47      1267



### SVC

In [41]:
Classifier = SVC()
Classifier.fit(vectors_train,y_train)

SVC()

In [42]:
y_pred = Classifier.predict(vectors_test)
y_pred

array(['REAL', 'FAKE', 'FAKE', ..., 'FAKE', 'REAL', 'FAKE'], dtype=object)

In [43]:
accuracy = accuracy_score(y_test,y_pred)

cm = confusion_matrix(y_test,y_pred)

report = classification_report(y_test,y_pred)
print('Accuracy:', accuracy, '\n',
      'Confusion Matrix:\n', cm, '\n',
      'Report:\n', report)

Accuracy: 0.9265982636148382 
 Confusion Matrix:
 [[577  38]
 [ 55 597]] 
 Report:
               precision    recall  f1-score   support

        FAKE       0.91      0.94      0.93       615
        REAL       0.94      0.92      0.93       652

    accuracy                           0.93      1267
   macro avg       0.93      0.93      0.93      1267
weighted avg       0.93      0.93      0.93      1267



### Decision Tree Classifier

In [44]:
Classifier = DecisionTreeClassifier()
Classifier.fit(vectors_train,y_train)

DecisionTreeClassifier()

In [45]:
y_pred = Classifier.predict(vectors_test)
y_pred

array(['REAL', 'FAKE', 'FAKE', ..., 'FAKE', 'FAKE', 'REAL'], dtype=object)

In [46]:
accuracy = accuracy_score(y_test,y_pred)

cm = confusion_matrix(y_test,y_pred)

report = classification_report(y_test,y_pred)
print('Accuracy:', accuracy, '\n',
      'Confusion Matrix:\n', cm, '\n',
      'Report:\n', report)

Accuracy: 0.8303078137332282 
 Confusion Matrix:
 [[497 118]
 [ 97 555]] 
 Report:
               precision    recall  f1-score   support

        FAKE       0.84      0.81      0.82       615
        REAL       0.82      0.85      0.84       652

    accuracy                           0.83      1267
   macro avg       0.83      0.83      0.83      1267
weighted avg       0.83      0.83      0.83      1267



### Random Forest Classifier

In [47]:
Classifier = RandomForestClassifier()
Classifier.fit(vectors_train,y_train)

RandomForestClassifier()

In [48]:
y_pred = Classifier.predict(vectors_test)
y_pred

array(['REAL', 'FAKE', 'FAKE', ..., 'FAKE', 'REAL', 'REAL'], dtype=object)

In [49]:
accuracy = accuracy_score(y_test,y_pred)

cm = confusion_matrix(y_test,y_pred)

report = classification_report(y_test,y_pred)
print('Accuracy:', accuracy, '\n',
      'Confusion Matrix:\n', cm, '\n',
      'Report:\n', report)

Accuracy: 0.9076558800315706 
 Confusion Matrix:
 [[547  68]
 [ 49 603]] 
 Report:
               precision    recall  f1-score   support

        FAKE       0.92      0.89      0.90       615
        REAL       0.90      0.92      0.91       652

    accuracy                           0.91      1267
   macro avg       0.91      0.91      0.91      1267
weighted avg       0.91      0.91      0.91      1267



## Over all the best model identified by Using --> TF-IDF Vectorizer Method -> PassiveAggressiveClassifier Algorithm gives 0.93% Accuracy